File: src/config/schema.ts
import { z } from "zod/v4";

export const COMMAND_IDS = [
  "window:create",
  "window:next",
  "window:previous",
  "window:close",
  "window:rename",
  "pane:split-horizontal",
  "pane:split-vertical",
  "pane:next",
  "pane:close",
  "pane:zoom",
  "pane:focus",
  "pane:focus-up",
  "pane:focus-down",
  "pane:focus-left",
  "pane:focus-right",
  "session:detach",
  "session:create",
  "session:list",
  "session:find",
  "session:rename",
  "server:kill",
  "command-palette",
  "keybindings:show",
  "copy-mode:enter",
  "notes:create",
  "notes:list",
] as const;

export const CommandIdSchema = z.enum(COMMAND_IDS);
export type CommandId = z.infer<typeof CommandIdSchema>;

export const KeybindingValueSchema = z.union([
  CommandIdSchema,
  z.object({
    command: CommandIdSchema,
    unless: z.array(z.string()).optional(),
  }),
]);
export type KeybindingValue = z.infer<typeof KeybindingValueSchema>;

export const STATUSBAR_MODULE_IDS = [
  "session",
  "windows",
  "datetime",
  "hostname",
  "user",
  "cwd",
  "git",
  "cpu",
  "ram",
  "battery",
  "network",
  "prefix",
  "pane-info",
] as const;

export const StatusBarModuleIdSchema = z.enum(STATUSBAR_MODULE_IDS);
export type StatusBarModuleId = z.infer<typeof StatusBarModuleIdSchema>;

export const StatusBarThemeSchema = z.object({
  bg: z.string().default("#1e1e2e"),
  fg: z.string().default("#cdd6f4"),
  active: z.string().default("#89b4fa"),
});

export const BorderThemeSchema = z.object({
  style: z.enum(["rounded", "sharp", "double", "none"]).default("rounded"),
  fg: z.string().default("#585b70"),
  activeFg: z.string().default("#89b4fa"),
});

export const ThemeSchema = z.object({
  statusBar: StatusBarThemeSchema.default(() => ({
    bg: "#1e1e2e",
    fg: "#cdd6f4",
    active: "#89b4fa",
  })),
  border: BorderThemeSchema.default(() => ({
    style: "rounded" as const,
    fg: "#585b70",
    activeFg: "#89b4fa",
  })),
});

export const StatusBarSeparatorSchema = z.object({
  style: z
    .enum(["powerline", "rounded", "flat", "arrow", "slant"])
    .default("powerline"),
  left: z.string().optional(),
  right: z.string().optional(),
});

export const StatusBarModuleConfigSchema = z
  .object({
    enabled: z.boolean().default(true),
    fg: z.string().optional(),
    bg: z.string().optional(),
  })
  .passthrough();

export const StatusBarConfigSchema = z.object({
  enabled: z.boolean().default(true),
  position: z.enum(["top", "bottom"]).default("bottom"),
  theme: z
    .enum([
      "catppuccin-mocha",
      "dracula",
      "nord",
      "tokyo-night",
      "gruvbox",
      "one-dark",
      "solarized",
      "custom",
    ])
    .default("catppuccin-mocha"),
  separator: StatusBarSeparatorSchema.default(() => ({
    style: "powerline" as const,
  })),
  icons: z.boolean().default(true),
  left: z
    .array(StatusBarModuleIdSchema)
    .default(["session", "windows"] as const),
  right: z
    .array(StatusBarModuleIdSchema)
    .default(["git", "cwd", "datetime"] as const),
  modules: z
    .record(z.string(), StatusBarModuleConfigSchema)
    .default(() => ({})),
  refreshInterval: z.number().default(1000),
  metricsInterval: z.number().default(5000),
});

export type StatusBarConfig = z.infer<typeof StatusBarConfigSchema>;

export const SessionsConfigSchema = z.object({
  autoSave: z.boolean().default(true),
  autoSaveInterval: z.number().default(30_000),
  autoRestore: z.boolean().default(true),
  savePath: z.string().default("~/.maxmux/sessions/"),
});

export const SessionListConfigSchema = z.object({
  mode: z.enum(["sidebar", "overlay"]).default("sidebar"),
  sidebarPosition: z.enum(["left", "right"]).default("left"),
  sidebarWidth: z.number().min(20).max(80).default(30),
});

export type SessionListConfig = z.infer<typeof SessionListConfigSchema>;

export const ConfigSchema = z.object({
  prefixKey: z.string().default("C-a"),
  prefixTimeout: z.number().default(0),
  historyLimit: z.number().min(0).max(100_000).default(10_000),
  shell: z.string().default(process.env.SHELL || "/bin/bash"),
  newPaneCwd: z.string().default("inherit"),
  switchToNewWindow: z.boolean().default(true),
  automaticRename: z.boolean().default(true),
  automaticRenameInterval: z.number().default(2000),
  mouse: z.boolean().default(true),
  showPrefixHelp: z.boolean().default(true),
  theme: ThemeSchema.default(() => ({
    statusBar: { bg: "#1e1e2e", fg: "#cdd6f4", active: "#89b4fa" },
    border: {
      style: "rounded" as const,
      fg: "#585b70",
      activeFg: "#89b4fa",
    },
  })),
  keybindings: z.record(z.string(), KeybindingValueSchema).default(() => ({})),
  globalKeybindings: z
    .record(z.string(), KeybindingValueSchema)
    .default(() => ({})),
  sessions: SessionsConfigSchema.default(() => ({
    autoSave: true,
    autoSaveInterval: 30_000,
    autoRestore: true,
    savePath: "~/.maxmux/sessions/",
  })),
  statusBar: StatusBarConfigSchema.default(() => ({
    enabled: true,
    position: "bottom" as const,
    theme: "catppuccin-mocha" as const,
    separator: { style: "powerline" as const },
    icons: true,
    left: ["session", "windows"] satisfies StatusBarModuleId[],
    right: ["git", "cwd", "datetime"] satisfies StatusBarModuleId[],
    modules: {},
    refreshInterval: 1000,
    metricsInterval: 5000,
  })),
  sessionList: SessionListConfigSchema.default(() => ({
    mode: "sidebar" as const,
    sidebarPosition: "left" as const,
    sidebarWidth: 30,
  })),
  plugins: z.array(z.any()).default(() => []),
  debug: z.boolean().default(false),
});

export type MaxMuxConfig = z.infer<typeof ConfigSchema>;

// -- Explicit input interfaces with JSDoc for LSP hover documentation --

/** Farbschema der Status Bar. Alle Werte sind Hex-Farbcodes. */
export interface StatusBarThemeInput {
  /** Hintergrundfarbe der Status Bar (Hex).
   * @default "#1e1e2e" */
  bg?: string;
  /** Textfarbe der Status Bar (Hex).
   * @default "#cdd6f4" */
  fg?: string;
  /** Akzentfarbe fuer aktive Elemente wie das aktive Fenster-Tab (Hex).
   * @default "#89b4fa" */
  active?: string;
}

/** Konfiguration fuer Pane-Raender. */
export interface BorderThemeInput {
  /** Stil der Rahmenlinien zwischen Panes.
   * - `"rounded"` — Abgerundete Ecken (Unicode Box-Drawing)
   * - `"sharp"` — Eckige Ecken
   * - `"double"` — Doppelte Linien
   * - `"none"` — Keine sichtbaren Rahmen
   * @default "rounded" */
  style?: "rounded" | "sharp" | "double" | "none";
  /** Farbe der Rahmenlinien fuer inaktive Panes (Hex).
   * @default "#585b70" */
  fg?: string;
  /** Farbe der Rahmenlinien fuer das aktive Pane (Hex).
   * @default "#89b4fa" */
  activeFg?: string;
}

/** Visuelles Gesamtthema — Farben fuer Status Bar und Rahmen. */
export interface ThemeInput {
  /** Farben der Status Bar. */
  statusBar?: StatusBarThemeInput;
  /** Rahmenlinien zwischen Panes. */
  border?: BorderThemeInput;
}

/** Konfiguration der Separator-Zeichen zwischen Status-Bar-Modulen. */
export interface StatusBarSeparatorInput {
  /** Visueller Stil der Trennzeichen.
   * - `"powerline"` — Powerline-Pfeile (erfordert Nerd Font)
   * - `"rounded"` — Abgerundete Trennzeichen
   * - `"flat"` — Flache Bloecke ohne Uebergang
   * - `"arrow"` — Schmale Pfeil-Trennzeichen
   * - `"slant"` — Schraege Trennzeichen
   * @default "powerline" */
  style?: "powerline" | "rounded" | "flat" | "arrow" | "slant";
  /** Eigenes Zeichen fuer den linken Separator. Ueberschreibt den Style-Default. */
  left?: string;
  /** Eigenes Zeichen fuer den rechten Separator. Ueberschreibt den Style-Default. */
  right?: string;
}

/** Konfiguration eines einzelnen Status-Bar-Moduls. */
export interface StatusBarModuleConfigInput {
  /** Ob dieses Modul angezeigt wird.
   * @default true */
  enabled?: boolean;
  /** Eigene Textfarbe fuer dieses Modul (Hex). Ueberschreibt die Theme-Farbe. */
  fg?: string;
  /** Eigene Hintergrundfarbe fuer dieses Modul (Hex). Ueberschreibt die Theme-Farbe. */
  bg?: string;
  /** Weitere modulspezifische Optionen (z.B. Format-Strings). */
  [key: string]: unknown;
}

/** Konfiguration der Status Bar am unteren oder oberen Bildschirmrand. */
export interface StatusBarInput {
  /** Ob die Status Bar sichtbar ist.
   * @default true */
  enabled?: boolean;
  /** Position der Status Bar.
   * @default "bottom" */
  position?: "top" | "bottom";
  /** Farbthema-Preset fuer die Status Bar. Bei `"custom"` werden die Farben
   * aus `theme.statusBar` verwendet.
   * @default "catppuccin-mocha" */
  theme?:
    | "catppuccin-mocha"
    | "dracula"
    | "nord"
    | "tokyo-night"
    | "gruvbox"
    | "one-dark"
    | "solarized"
    | "custom";
  /** Trennzeichen zwischen den Status-Bar-Modulen. */
  separator?: StatusBarSeparatorInput;
  /** Nerd-Font-Icons in Status-Bar-Modulen anzeigen. Erfordert einen
   * Nerd Font als Terminal-Schrift.
   * @default true */
  icons?: boolean;
  /** Module auf der linken Seite der Status Bar, in Reihenfolge.
   * @default ["session", "windows"] */
  left?: StatusBarModuleId[];
  /** Module auf der rechten Seite der Status Bar, in Reihenfolge.
   * @default ["git", "cwd", "datetime"] */
  right?: StatusBarModuleId[];
  /** Individuelle Konfiguration pro Modul. Key = Modul-ID. */
  modules?: Record<string, StatusBarModuleConfigInput>;
  /** Aktualisierungsintervall der Status Bar in Millisekunden.
   * @default 1000 */
  refreshInterval?: number;
  /** Intervall in ms fuer das Abfragen von Systemmetriken (CPU, RAM, etc).
   * @default 5000 */
  metricsInterval?: number;
}

/** Session-Persistenz — automatisches Speichern und Wiederherstellen. */
export interface SessionsInput {
  /** Sessions periodisch automatisch speichern.
   * @default true */
  autoSave?: boolean;
  /** Intervall in ms zwischen Auto-Saves.
   * @default 30000 */
  autoSaveInterval?: number;
  /** Gespeicherte Sessions beim Serverstart automatisch wiederherstellen.
   * @default true */
  autoRestore?: boolean;
  /** Verzeichnis fuer Session-Dateien. Wird automatisch erstellt.
   * @default "~/.maxmux/sessions/" */
  savePath?: string;
}

/** Konfiguration der Session-Liste/Picker-Ansicht. */
export interface SessionListInput {
  /** Anzeigemodus der Session-Liste.
   * - `"sidebar"` — Feste Seitenleiste (bleibt sichtbar)
   * - `"overlay"` — Schwebendes Popup ueber dem Terminal
   * @default "sidebar" */
  mode?: "sidebar" | "overlay";
  /** Seite der Seitenleiste (nur bei `mode: "sidebar"`).
   * @default "left" */
  sidebarPosition?: "left" | "right";
  /** Breite der Seitenleiste in Spalten (nur bei `mode: "sidebar"`).
   * Min: 20, Max: 80.
   * @default 30 */
  sidebarWidth?: number;
}

/**
 * MaxMux-Konfiguration. Alle Felder sind optional — nicht angegebene
 * Felder verwenden die Standardwerte. Verwende {@link defineConfig} fuer
 * typsichere Konfiguration mit Validierung.
 *
 * @example
 * ```ts
 * // maxmux.config.ts
 * import { defineConfig } from "maxmux";
 *
 * export default defineConfig({
 *   prefixKey: "C-b",
 *   theme: {
 *     border: { style: "sharp" },
 *   },
 * });
 * ```
 */
export interface MaxMuxConfigInput {
  /** Prefix-Taste, die den Keybinding-Modus aktiviert.
   * Format: `"C-x"` fuer Ctrl+x. Nach dem Druecken dieser Taste wird
   * der naechste Tastendruck als Keybinding interpretiert.
   * @default "C-a" */
  prefixKey?: string;
  /** Timeout in ms nach dem Druecken der Prefix-Taste, bevor der
   * Prefix-Modus automatisch zurueckgesetzt wird.
   * `0` = kein Timeout (wartet unbegrenzt auf den naechsten Tastendruck).
   * @default 0 */
  prefixTimeout?: number;
  /** Maximale Anzahl an Zeilen im Scrollback-Buffer pro Pane.
   * Hoehere Werte verbrauchen mehr Speicher.
   * Min: 0, Max: 100.000.
   * @default 10000 */
  historyLimit?: number;
  /** Shell-Programm, das fuer neue Panes gestartet wird.
   * Verwendet standardmaessig die `$SHELL`-Umgebungsvariable.
   * @default $SHELL oder "/bin/bash" */
  shell?: string;
  /** Arbeitsverzeichnis fuer neue Panes.
   * - `"inherit"` — Verzeichnis des aktiven Panes uebernehmen
   * - Ein absoluter Pfad (z.B. `"/home/user/projects"`) — immer dieses Verzeichnis verwenden
   * @default "inherit" */
  newPaneCwd?: string;
  /** Automatisch zum neu erstellten Fenster wechseln.
   * @default true */
  switchToNewWindow?: boolean;
  /** Fenster automatisch nach dem laufenden Prozess umbenennen.
   * @default true */
  automaticRename?: boolean;
  /** Intervall in ms fuer die Prozessnamens-Pruefung bei automatischer Umbenennung.
   * @default 2000 */
  automaticRenameInterval?: number;
  /** Maus-Unterstuetzung aktivieren (Klick zum Fokussieren, Scrollen).
   * @default true */
  mouse?: boolean;
  /** Keybinding-Hilfe-Popup nach Druecken der Prefix-Taste anzeigen.
   * Zeigt eine kurze Uebersicht der verfuegbaren Tastenkuerzel.
   * @default true */
  showPrefixHelp?: boolean;
  /** Visuelles Theme — Farben fuer Rahmen und Status Bar. */
  theme?: ThemeInput;
  /** Keybindings im Prefix-Modus. Tasten, die nach der Prefix-Taste
   * gedrueckt werden. Werden mit den Standard-Keybindings gemerged
   * (eigene Eintraege ueberschreiben Defaults).
   *
   * Keys: einzelne Zeichen (`"c"`, `"%"`) oder Spezialtasten (`"Up"`, `"Down"`).
   * Values: Command-ID (`"window:create"`) oder Objekt mit `command` und `unless`.
   *
   * @example
   * ```ts
   * keybindings: {
   *   "c": "window:create",
   *   "|": "pane:split-horizontal",
   *   "-": "pane:split-vertical",
   * }
   * ``` */
  keybindings?: Record<string, KeybindingValue>;
  /** Globale Keybindings — feuern sofort ohne Prefix-Taste.
   *
   * **Achtung:** Globale Bindings fangen den Tastendruck ab, bevor er an
   * die Shell/das laufende Programm weitergeleitet wird. Verwende nur
   * Tastenkombinationen, die du nicht im Terminal brauchst.
   *
   * Keys: Ctrl-Kombinationen (`"C-h"`, `"C-j"`), Spezialtasten, oder Zeichen.
   * Values: Command-ID oder Objekt mit `command` und `unless`.
   *
   * @default {} (keine globalen Bindings)
   * @example
   * ```ts
   * globalKeybindings: {
   *   "C-h": "pane:focus-left",
   *   "C-l": "pane:focus-right",
   * }
   * ``` */
  globalKeybindings?: Record<string, KeybindingValue>;
  /** Session-Persistenz — automatisches Speichern und Wiederherstellen. */
  sessions?: SessionsInput;
  /** Status Bar am Bildschirmrand — Module, Position und Farbthema. */
  statusBar?: StatusBarInput;
  /** Session-Liste/Picker — Darstellungsmodus und Seitenleisten-Optionen. */
  sessionList?: SessionListInput;
  /** Plugins, die beim Start geladen werden. Jedes Plugin stellt eine
   * `setup(ctx)`-Funktion bereit, die Zugriff auf Commands, Keybindings
   * und Event-Hooks erhaelt. */
  plugins?: unknown[];
  /** Debug-Logging in Datei aktivieren. Wenn `true`, werden interne
   * Ereignisse (Server, Client, PTY) in `~/.maxmux/debug.log` protokolliert.
   * Nuetzlich zur Fehleranalyse. Im Normalbetrieb deaktiviert lassen.
   * @default false */
  debug?: boolean;
}

/**
 * Validiert und vervollstaendigt die MaxMux-Konfiguration.
 * Fehlende Felder werden mit Standardwerten aufgefuellt.
 * Wirft einen Fehler bei ungueltigen Werten.
 *
 * @example
 * ```ts
 * // maxmux.config.ts
 * import { defineConfig } from "maxmux";
 *
 * export default defineConfig({
 *   prefixKey: "C-b",
 *   mouse: true,
 *   theme: {
 *     border: { style: "sharp" },
 *   },
 * });
 * ```
 */
export function defineConfig(config: MaxMuxConfigInput): MaxMuxConfig {
  return ConfigSchema.parse(config);
}

File: rust/crates/maxmux-config/src/schema.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::env;

// ---------------------------------------------------------------------------
// Enums
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
#[derive(Default)]
pub enum BorderStyle {
    #[default]
    Rounded,
    Sharp,
    Double,
    None,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
#[derive(Default)]
pub enum LineStyle {
    #[default]
    Solid,
    Dashed,
    Dotted,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
#[derive(Default)]
pub enum Position {
    Top,
    #[default]
    Bottom,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
#[derive(Default)]
pub enum SeparatorStyle {
    #[default]
    Powerline,
    Rounded,
    Flat,
    Arrow,
    Slant,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
#[derive(Default)]
pub enum SessionListMode {
    #[default]
    Sidebar,
    Overlay,
}

// ---------------------------------------------------------------------------
// KeybindingValue — either a plain command string or { command, unless }
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(untagged)]
pub enum KeybindingValue {
    Command(String),
    Conditional {
        command: String,
        #[serde(default)]
        unless: Vec<String>,
    },
}

// ---------------------------------------------------------------------------
// Theme structs
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(default)]
pub struct StatusBarThemeColors {
    pub bg: String,
    pub fg: String,
    pub active: String,
}

impl Default for StatusBarThemeColors {
    fn default() -> Self {
        Self {
            bg: "#1e1e2e".into(),
            fg: "#cdd6f4".into(),
            active: "#89b4fa".into(),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(default)]
pub struct BorderThemeConfig {
    pub style: BorderStyle,
    pub line_style: LineStyle,
    pub fg: String,
    pub active_fg: String,
}

impl Default for BorderThemeConfig {
    fn default() -> Self {
        Self {
            style: BorderStyle::Rounded,
            line_style: LineStyle::Solid,
            fg: "#585b70".into(),
            active_fg: "#89b4fa".into(),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(default)]
#[derive(Default)]
pub struct ThemeConfig {
    pub status_bar: StatusBarThemeColors,
    pub border: BorderThemeConfig,
}

// ---------------------------------------------------------------------------
// Status bar
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(default)]
pub struct SeparatorConfig {
    pub style: SeparatorStyle,
    pub left: Option<String>,
    pub right: Option<String>,
}

impl Default for SeparatorConfig {
    fn default() -> Self {
        Self {
            style: SeparatorStyle::Powerline,
            left: None,
            right: None,
        }
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(default)]
pub struct StatusBarConfig {
    pub enabled: bool,
    pub position: Position,
    pub theme: String,
    pub separator: SeparatorConfig,
    pub icons: bool,
    pub left: Vec<String>,
    pub right: Vec<String>,
    pub modules: HashMap<String, toml::Value>,
    pub refresh_interval: u64,
    pub metrics_interval: u64,
}

impl Default for StatusBarConfig {
    fn default() -> Self {
        Self {
            enabled: true,
            position: Position::Bottom,
            theme: "catppuccin-mocha".into(),
            separator: SeparatorConfig::default(),
            icons: true,
            left: vec!["session".into(), "windows".into()],
            right: vec!["git".into(), "cwd".into(), "datetime".into()],
            modules: HashMap::new(),
            refresh_interval: 1000,
            metrics_interval: 5000,
        }
    }
}

// ---------------------------------------------------------------------------
// Sessions
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(default)]
pub struct SessionsConfig {
    pub auto_save: bool,
    pub auto_save_interval: u64,
    pub auto_restore: bool,
    pub save_path: String,
}

impl Default for SessionsConfig {
    fn default() -> Self {
        Self {
            auto_save: true,
            auto_save_interval: 30_000,
            auto_restore: true,
            save_path: "~/.maxmux/sessions/".into(),
        }
    }
}

// ---------------------------------------------------------------------------
// Session list
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(default)]
pub struct SessionListConfig {
    pub mode: SessionListMode,
    pub sidebar_position: String,
    pub sidebar_width: u16,
}

impl Default for SessionListConfig {
    fn default() -> Self {
        Self {
            mode: SessionListMode::Sidebar,
            sidebar_position: "left".into(),
            sidebar_width: 30,
        }
    }
}

// ---------------------------------------------------------------------------
// Root config
// ---------------------------------------------------------------------------

fn default_shell() -> String {
    env::var("SHELL").unwrap_or_else(|_| "/bin/bash".into())
}

/// Returns the full set of default prefix-mode keybindings.
pub fn default_keybindings() -> HashMap<String, KeybindingValue> {
    let bindings: &[(&str, &str)] = &[
        ("c", "window:create"),
        ("n", "window:next"),
        ("p", "window:previous"),
        ("&", "window:close"),
        (",", "window:rename"),
        ("|", "pane:split-horizontal"),
        ("-", "pane:split-vertical"),
        ("o", "pane:next"),
        ("x", "pane:close"),
        ("z", "pane:zoom"),
        ("Up", "pane:focus-up"),
        ("Down", "pane:focus-down"),
        ("Left", "pane:focus-left"),
        ("Right", "pane:focus-right"),
        ("d", "session:detach"),
        ("$", "session:rename"),
        ("s", "session:find"),
        (":", "command-palette"),
        ("?", "keybindings:show"),
        ("[", "copy-mode:enter"),
        ("N", "notes:create"),
        ("L", "notes:list"),
    ];
    bindings
        .iter()
        .map(|(k, v)| ((*k).to_string(), KeybindingValue::Command((*v).to_string())))
        .collect()
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(default)]
pub struct MaxmuxConfig {
    pub prefix_key: String,
    pub prefix_timeout: u64,
    pub history_limit: u32,
    pub shell: String,
    pub new_pane_cwd: String,
    pub switch_to_new_window: bool,
    pub automatic_rename: bool,
    pub automatic_rename_interval: u64,
    pub mouse: bool,
    pub show_prefix_help: bool,
    pub debug: bool,
    pub theme: ThemeConfig,
    pub keybindings: HashMap<String, KeybindingValue>,
    pub global_keybindings: HashMap<String, KeybindingValue>,
    pub status_bar: StatusBarConfig,
    pub sessions: SessionsConfig,
    pub session_list: SessionListConfig,
    pub plugins: Vec<toml::Value>,
}

impl Default for MaxmuxConfig {
    fn default() -> Self {
        Self {
            prefix_key: "C-a".into(),
            prefix_timeout: 0,
            history_limit: 10_000,
            shell: default_shell(),
            new_pane_cwd: "inherit".into(),
            switch_to_new_window: true,
            automatic_rename: true,
            automatic_rename_interval: 2000,
            mouse: true,
            show_prefix_help: true,
            debug: false,
            theme: ThemeConfig::default(),
            keybindings: default_keybindings(),
            global_keybindings: HashMap::new(),
            status_bar: StatusBarConfig::default(),
            sessions: SessionsConfig::default(),
            session_list: SessionListConfig::default(),
            plugins: Vec::new(),
        }
    }
}

// ---------------------------------------------------------------------------
// Validation
// ---------------------------------------------------------------------------

#[derive(Debug, thiserror::Error)]
pub enum ConfigValidationError {
    #[error("history_limit must be between 0 and 100000, got {0}")]
    HistoryLimitOutOfRange(u32),
    #[error("sidebar_width must be between 20 and 80, got {0}")]
    SidebarWidthOutOfRange(u16),
}

impl MaxmuxConfig {
    /// Validate config values that cannot be expressed via serde alone.
    pub fn validate(&self) -> Result<(), ConfigValidationError> {
        if self.history_limit > 100_000 {
            return Err(ConfigValidationError::HistoryLimitOutOfRange(
                self.history_limit,
            ));
        }
        if self.session_list.sidebar_width < 20 || self.session_list.sidebar_width > 80 {
            return Err(ConfigValidationError::SidebarWidthOutOfRange(
                self.session_list.sidebar_width,
            ));
        }
        Ok(())
    }
}
